internal/pkg/override/patch_test.go

```diff
@@ -0,0 +1,452 @@
+- op: replace
+  path: /Resources/TaskDef/Description
+  value: jkl;`,
+			expected: `
+Resources:
+  TaskDef:
+    Type: AWS::ECS::TaskDefinition
+    Description: jkl;`,
+		},
+		"replace map with scalar": {
+			yaml: `
+Resources:
+  List:
+    - asdf
+    - key: value
+      key2: value2
+    - - very list
+      - many item`,
+			overrides: `
+- op: replace
+  path: /Resources/List/1
+  value: jkl;`,
+			expected: `
+Resources:
+  List:
+    - asdf
+    - jkl;
+    - - very list
+      - many item`,
+		},
+		"works with special characters": {
+			yaml: `
+Resources:
+  key:
+    key~with/weirdchars/: old`,
+			overrides: `
+- op: replace
+  path: /Resources/key/key~0with~1weirdchars~1
+  value: new`,
+			expected: `
+Resources:
+  key:
+    key~with/weirdchars/: new`,
+		},
+		"add works with doc selector": {
+			yaml: `
+Resources:
+  key: value`,
+			overrides: `
+- op: add
+  path: ""
+  value:
+    a: aaa
+    b: bbb`,
+			expected: `
+a: aaa
+b: bbb`,
+		},
+		"replace works with doc selector": {
+			yaml: `
+Resources:
+  key: value`,
+			overrides: `
+- op: replace
+  path: ""
+  value:
+    a: aaa
+    b: bbb`,
+			expected: `
+a: aaa
+b: bbb`,
+		},
+		"remove works with doc selector": {
+			yaml: `
+Resources:
+  key: value`,
+			overrides: `
+- op: remove
+  path: ""`,
+			expected: ``,
+		},
+		"empty string key works": {
+			yaml: `
+key: asdf
+"": old`,
+			overrides: `
+- op: replace
+  path: /
+  value: new`,
+			expected: `
+key: asdf
+"": new`,
+		},
+		"error on invalid patch file format": {
+			overrides: `
+op: add
+path: /
+value: new`,
+			expectedErr: `file at "/cfn.patches.yaml" does not conform to the YAML patch document schema: yaml: unmarshal errors:
+  line 1: cannot unmarshal !!map into []override.yamlPatch`,
+		},
+		"error on unsupported operation": {
+			overrides: `
+- op: unsupported
+  path: /
+  value: new`,
+			expectedErr: `unsupported operation "unsupported". supported operations are "add", "remove", and "replace".`,
+		},
+		"error in map following path": {
+			yaml: `
+a:
+  b:
+    - c
+    - d`,
+			overrides: `
+- op: replace
+  path: /a/e/c
+  value: val`,
+			expectedErr: `unable to apply "replace" patch: key "/a": "e" not found in map`,
+		},
+		"error out of bounds sequence following path": {
+			yaml: `
+a:
+  b:
+    - c
+    - d`,
+			overrides: `
+- op: add
+  path: /a/b/3
+  value: val`,
+			expectedErr: `unable to apply "add" patch: key "/a/b": invalid index 3 for sequence of length 2`,
+		},
+		"error invalid index sequence following path": {
+			yaml: `
+a:
+  b:
+    - c
+    - d`,
+			overrides: `
+- op: add
+  path: /a/b/e
+  value: val`,
+			expectedErr: `unable to apply "add" patch: key "/a/b": expected index in sequence, got "e"`,
+		},
+		"error targeting scalar while following path": {
+			yaml: `
+a:
+  b:
+    - c
+    - d`,
+			overrides: `
+- op: add
+  path: /a/b/1/e
+  value: val`,
+			expectedErr: `unable to apply "add" patch: key "/a/b/1": invalid node type 0x8`,
```

Comment:
😅  the node type `0x8` won't make sense to users, can we add a `nodeKindStringer` type to make the error more friendly?
```go
type nodeKindStringer yaml.Kind

// String implements fmt.Stringer.
func (k nodeKindStringer) String() string {
   switch k {
   case yaml.DocumentNode:
      return "document"
   case yaml.MappingNode:
      return "mapping"
   // ...
   }
}
```

and in `followJSONPointerHelper`:
```go
return fmt.Errorf("key %q: invalid node type %q", strings.Join(traversed, jsonPointerSeparator), nodeKindStringer(node.Kind))
```